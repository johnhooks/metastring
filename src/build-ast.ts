import { Parser } from "acorn";
import { BaseNode, Program } from "estree";

import { MetaParserPlugin } from "./meta-parser-plugin.js";

/**
 * Build AST from Markdown code fence metadata.
 *
 * Uses `acorn` to do the heavy lifting, overriding parsing features that aren't
 * necessary for this project.
 * @public
 */
export function buildAst(input: string) {
	const metaParser = Parser.extend(MetaParserPlugin);
	const ast = metaParser.parse(input, { ecmaVersion: 2020 });
	return ast as BaseNode as Program;
}
