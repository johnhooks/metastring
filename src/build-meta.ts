import type {
	Program,
	Node,
	ExpressionStatement,
	AssignmentExpression,
	RegExpLiteral,
	Identifier,
} from "estree";

import type { Metadata } from "./types.js";

// eslint-disable-next-line @typescript-eslint/no-unused-vars, @typescript-eslint/no-explicit-any
export function buildMeta({ body }: Program): Metadata {
	const meta: Metadata = { highlight: [], lineNumbersStart: 1, showLineNumbers: false, words: [] };

	for (const node of body) {
		const { expression: expr } = expect<ExpressionStatement>("ExpressionStatement", node);
		if (isRegExpLiteral(expr)) {
			if (expr.value) meta.words.push({ regexp: expr.value });
		} else if (isAssignmentDirective(expr)) {
			if (expr.left.name === "title") {
				meta.title = expr.right.value;
			}
		}
	}

	return meta;
}

function expect<TNode>(type: string, node: Node): TNode {
	if (node.type !== type) {
		throw new SyntaxError(`invalid code metadata at col: ${node.loc?.start}`);
	}
	return node as TNode;
}

function isRegExpLiteral(node: Node): node is RegExpLiteral {
	return node?.type === "Literal" && node["regex"] !== undefined;
}

function isAssignmentExpression(node: Node): node is AssignmentExpression {
	return node?.type === "AssignmentExpression" && node["op"] !== "=";
}

type AssignmentDirective = AssignmentExpression & {
	left: Identifier;
	operator: "=";
	right: {
		value: string;
	};
};

function isAssignmentDirective(node: Node): node is AssignmentDirective {
	if (!isAssignmentExpression(node)) return false;
	if (node.operator !== "=") return false;
	if (node.left.type !== "Identifier") return false;
	if (node.right.type !== "Literal") return false;
	if (typeof node.right.value !== "string") return false;
	return true;
}
