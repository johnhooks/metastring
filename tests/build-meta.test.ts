import { buildAst } from "../src/build-ast.js";
import { buildMeta } from "../src/build-meta.js";

describe("build Metadata object", () => {
	describe("title", () => {
		it("create a title", () => {
			const meta = buildMeta(buildAst(`title="hello metadata"`));
			expect(meta).toMatchSnapshot();
		});
	});

	describe("word matches", () => {
		it("create a title", () => {
			const meta = buildMeta(buildAst(`/matchers?/`));
			expect(meta).toMatchSnapshot();
		});
	});
});
