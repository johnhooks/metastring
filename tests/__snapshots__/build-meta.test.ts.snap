// Vitest Snapshot v1

exports[`build Metadata object > title > create a title 1`] = `
{
  "highlight": [],
  "lineNumbersStart": 1,
  "showLineNumbers": false,
  "title": "hello metadata",
  "words": [],
}
`;

exports[`build Metadata object > word matches > create a title 1`] = `
{
  "highlight": [],
  "lineNumbersStart": 1,
  "showLineNumbers": false,
  "words": [
    {
      "regexp": /matchers\\?/,
    },
  ],
}
`;
